File: junk/string/lcs-memo.cpp
#include <iostream>
#include <string>
#include <map>

using namespace std;

string LCS(string S, string T, map < pair<string, string> , string > &memo)
{
  if (S.size() == 0 || T.size() == 0)
    return "";

  pair<string, string> idx = make_pair(S,T);
  map < pair<string, string> ,  string >::iterator it = memo.find(idx);
  if (it != memo.end())
    return it->second;

  string ret;

  if (S[0] == T[0])
    ret = S[0] + LCS(S.substr(1),  T.substr(1), memo);
  else {
    string s1 = LCS(S.substr(1), T, memo);
    string s2 = LCS(S, T.substr(1), memo);
    ret = s1.size() > s2.size() ? s1 : s2;
  }

  memo[idx] = ret;
  // Uncomment to see how the cache "memo" is filled out
  //cerr << S << ", " << T << " => " << ret << endl;

  return ret;
}

int
main(int argc, char *argv[])
{
  map < pair<string, string> ,  string > memo;
  cout << "LCS:" << LCS("abcdefghijklmnzyxwvutsrqpo",
                        "opqrstuvwxyzabcdefghijklmn", memo) << endl;
  return 0;
}

File: junk/string/lcs-memo.py
#!/usr/bin/python

def LCS(S, T, memo):

    if not len(S) or not len(T):
        return ''

    if memo.has_key((S,T)):
        return memo[(S,T)]

    if S[0] == T[0]:
        ret = S[0] + LCS(S[1:],  T[1:], memo)
    else:
        s1 = LCS(S[1:], T, memo)
        s2 = LCS(S, T[1:], memo)
        ret = s1 if len(s1) > len(s2) else s2

    memo[(S,T)] = ret
    #Uncomment if you want to see how the cache is created.
    #print (S,T), '=', ret

    return ret

print LCS('abcdefghijklmnzyxwvutsrqpo', 'opqrstuvwxyzabcdefghijklmn', {})
